### PR TITLE
feat(storage): restore soft deleted object using restore token for HNS buckets

### PR DESCRIPTION
**Pull request by shubhangi-google**
Add support for restore token to soft deleted from HNS bucket.

Operation Supported:

- If a restore token is provided as an input to restore_file method along with flename and generation token, it will restore seamlessly.
- If restore token is provided, the method calls works as earlier filename and generation token.
- Invalid restore tokens will result in a 404 Not Found error from the Storage Transfer API.

```ruby
 @example
require "google/cloud/storage"
storage = Google::Cloud::Storage.new
bucket = storage.bucket "my-hns-bucket"
bucket.restore_file "path/of/file", <generation-of-the-file>, "<restore_token>"
```

### CODE REVIEW

File: google-cloud-storage/lib/google/cloud/storage/bucket.rb [hunk 1]
```diff
@@ -1843,6 +1843,8 @@ def create_file file,
         #   Name of the file.
         # @param [Fixnum] generation
         #   Selects a specific revision of this object.
+        # @param [String] restore_token
```
Comment (bajajneha27) [inline google-cloud-storage/lib/google/cloud/storage/bucket.rb:1846]: I believe we need to `restore_token` for both `get` and `list` objects operations. Refer to go/hns-bucket-softdelete-support-client-request

**Review by bajajneha27 — CHANGES_REQUESTED** [review]

File: google-cloud-storage/lib/google/cloud/storage/bucket.rb (hunk 1 above)
Comment (shubhangi-google) [reply google-cloud-storage/lib/google/cloud/storage/bucket.rb:1846]: Hi @bajajneha27 this is already handled we are able to fetch restore token in get and list request
```ruby
@example
bucket.file "try_sub_a/flowers.jpeg", soft_deleted: true, generation: 1729579373155129
=> 
#<Google::Cloud::Storage::File:0x00007f1e2591a7b8
 @gapi=
  #<Google::Apis::StorageV1::Object:0x00007f1e267afdc8
   @bucket="shubhangi_hns5",
   @content_type="image/jpeg",
   @crc32c="1Nykzg==",
   @etag="CLmml6GxoYkDEAE=",
   @generation=1729579373155129,
   @hard_delete_time=#<DateTime: 2024-11-06T12:43:53+00:00 ((2460621j,45833s,248000000n),+0s,2299161j)>,
   @id="shubhangi_hns5/try_sub_a/flowers.jpeg/1729579373155129",
   @kind="storage#object",
   @md5_hash="wISfTGSx0F0r5A+2uS5Hxw==",
   @metageneration=1,
   @name="try_sub_a/flowers.jpeg",
   @restore_token="d3e6275a-46c1-4d65-9f8f-4774e3bf421c",
   @self_link="https://www.googleapis.com/storage/v1/b/shubhangi_hns5/o/try_sub_a%2Fflowers.jpeg",
   @size=11899,
   @soft_delete_time=#<DateTime: 2024-10-30T12:43:53+00:00 ((2460614j,45833s,248000000n),+0s,2299161j)>,
   @storage_class="STANDARD",
   @time_created=#<DateTime: 2024-10-22T06:42:53+00:00 ((2460606j,24173s,245000000n),+0s,2299161j)>,
   @time_storage_class_updated=#<DateTime: 2024-10-22T06:42:53+00:00 ((2460606j,24173s,245000000n),+0s,2299161j)>,
   @updated=#<DateTime: 2024-10-22T06:42:53+00:00 ((2460606j,24173s,245000000n),+0s,2299161j)>>,
 @service=Google::Cloud::Storage::Service(storage-sdk-vendor),

```

Comment (bajajneha27) [reply google-cloud-storage/lib/google/cloud/storage/bucket.rb:1846]: We need to make sure that `file.restore_token` returns the token. We also need to make sure same thing happens for list operation.

Comment (shubhangi-google) [reply google-cloud-storage/lib/google/cloud/storage/bucket.rb:1846]: will add the check in unit test

Comment (bajajneha27) [reply google-cloud-storage/lib/google/cloud/storage/bucket.rb:1846]: Thank you!

**Comment by JesseLovelace** [comment]
We decided not to add this feature to the client libraries, and to direct customers to use the CLI or console instead

**Pull request closed without merging**